nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -2,148 +2,245 @@ pragma solidity ^0.5.3;
-        uint16 _lockedPeriods
+        uint256 _boostingRefund,
+        uint256 _lockingDuration
     )
         public
     {
-        require(_token.totalSupply() > 0 &&
+        uint256 totalSupply = _token.totalSupply();
+        require(totalSupply > 0 &&
             _escrow.secondsPerPeriod() > 0 &&
+            _router.target().isContract() &&
             _endBidDate > _startBidDate &&
             _endBidDate > block.timestamp &&
-            _depositRate > 0 &&
-            _refundRate > 0 &&
-            _lockedPeriods >= _escrow.minLockedPeriods());
+            _boostingRefund > 0 &&
+            _lockingDuration > 0);
+        // worst case for `ethToWork()` and `workToETH()`,
+        // when ethSupply == MAX_ETH_SUPPLY and tokenSupply == totalSupply
+        require(MAX_ETH_SUPPLY * totalSupply * SLOWING_REFUND / MAX_ETH_SUPPLY / totalSupply == SLOWING_REFUND &&
+            MAX_ETH_SUPPLY * totalSupply * _boostingRefund / MAX_ETH_SUPPLY / totalSupply == _boostingRefund);
+
         token = _token;
         escrow = _escrow;
+        router = _router;
         startBidDate = _startBidDate;
         endBidDate = _endBidDate;
-        minAllowableLockedTokens = _escrow.minAllowableLockedTokens();
-        maxAllowableLockedTokens = _escrow.maxAllowableLockedTokens();
-        depositRate = _depositRate;
-        refundRate = _refundRate;
-        lockedPeriods = _lockedPeriods;
+        boostingRefund = _boostingRefund;
+        lockingDuration = _lockingDuration;
+    }
+
+    /**
+    * @notice Deposit tokens to contract
+    * @param _value Amount of tokens to transfer
+    **/
+    function tokenDeposit(uint256 _value) external {
+        require(block.timestamp <= endBidDate, "Can't deposit more tokens after end of bidding");
+        token.safeTransferFrom(msg.sender, address(this), _value);
+        tokenSupply += _value;
+        emit Deposited(msg.sender, _value);
+    }
+
+    /**
+    * @notice Calculate amount of tokens that will be get for specified amount of ETH
+    * @dev This value will be fixed only after end of bidding
+    **/
+    function ethToTokens(uint256 _ethAmount) public view returns (uint256) {
+        return _ethAmount.mul(tokenSupply).div(ethSupply);
+    }
+
+    /**
+    * @notice Calculate amount of work that need to be done to refund specified amount of ETH
+    * @dev This value will be fixed only after end of bidding
+    **/
+    function ethToWork(uint256 _ethAmount) public view returns (uint256) {
+        return _ethAmount.mul(tokenSupply).mul(SLOWING_REFUND).divCeil(ethSupply.mul(boostingRefund));
+    }
+
+    /**
+    * @notice Calculate amount of ETH that will be refund for completing specified amount of work
+    * @dev This value will be fixed only after end of bidding
+    **/
+    function workToETH(uint256 _completedWork) public view returns (uint256) {
+        return _completedWork.mul(ethSupply).mul(boostingRefund).div(tokenSupply.mul(SLOWING_REFUND));
```

Comment:
Say for sake of argument that `SLOWING_REFUND` = `boostingRefund` – this would mean stakers must earn the exact same amount in inflationary rewards as their initial stake in order to reclaim all their locked ETH.

That sounds ok, except that the time it will take to get to reach this amount can vary enormously and, crucially, independently of the decisions of that staker. Let's assume an annual base inflation rate of 4% and a staker who locks their tokens to receive maximum rewards (i.e. duration >> 12 months), and restakes 100% of those rewards. This staker would face the following variance in the number of months ('#M') to reclaim the original ETH they locked in the Worklock, based on the participation rate ('PR'):
PR: 30% , #M: 7
PR: 50% , #M: 12
PR: 70% , #M: 20
PR: 90% , #M: 32
This isn't ideal, because the opportunity cost of locking up ETH for 1 year is very different to 2.5 years. The ETH price has dropped by 90% in 1 year before. 

If we want to give Worklock participants greater control/predictability over getting their ETH back, then this function should adjust based on the participation rate. In other words, if the participation rate goes up, then so should `boostingRefund`.